`Tutorial groups`: Allow instructors to create tutorial groups

stefanwaldhauser opened the pull request:
This feature is currently hidden based on a feature toggle

<!-- Thanks for contributing to Artemis! Before you submit your pull request, please make sure to check the following boxes by putting an x in the [ ] (don't: [x ], [ x], do: [x]) -->
<!-- If your pull request is not ready for review yet, create a draft pull request! -->

### Checklist
#### General
<!-- You only need to choose one of the first two check items: Generally, test on the test servers. -->
<!-- If it's only a small change, testing it locally is acceptable and you may remove the first checkmark. If you are unsure, please test on the test servers. -->
- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.ase.in.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Server
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I added `@PreAuthorize` and checked the course groups for all new REST Calls (security).
- [x] I implemented the changes with a good performance and prevented too many database calls.
- [x] I documented the Java code using JavaDoc style.
#### Client
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client/) and ensured that the layout is responsive.
- [x] Following the [theming guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client-tests/).
- [x] I added `authorities` to all new routes and checked the course groups for displaying navigation elements (links, buttons).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context
- First important step for the management of tutorial groups in Artemis : Basic entity and related CRUD(Create,Read,Update,Delete) functionality
- **Functionality included in this PR**:

1. Create,Update,Delete Tutorial Groups with basic properties like title, teaching assistant and so on
2. Registered Students management + import / export csv

**Remaining functionality like Session Management (When does a tutorial group happen?) will be done in further PRs**

### Description
- Removed existing TutorGroup entity and database tables
- Added new TutorialGroup entity and database tables
- Implemented CRUD functionality for TutorialGroups
- **Made the course group component reusable for the management of registered students of tutorial groups**

### Steps for Testing
<!-- Please describe in detail how the reviewer can test your changes. -->
Prerequisites:
- 1 Instructor
- 2 Feature Toggle for TutorialGroups activated

**Testing is pretty simple. Just create some tutorial groups, edit the information, delete them and register/deregister some students**


**Important: Because I reused the component for adding students, tutors and so on to courses for the management of registered students for a tutorial group it would be wise to test if adding students, tutors and so on to courses also still works as expected!**

### Review Progress
<!-- Each Pull Request should be reviewed by at least two other developers. The code as well as the functionality (= manual test) needs to be reviewed. -->
<!-- The reviewer or author check the following boxes depending on what was reviewed or tested. All boxes should be checked before merge. -->
<!-- You can add additional checkboxes if it makes sense to only review parts of the code or functionality. -->
<!-- When changes are pushed, uncheck the affected boxes. (Not all changes require full re-reviews.) -->

#### Code Review
- [x] Review 1
- [x] Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Test Coverage
<!-- Please add the test coverages for all changed files here. You can see this when executing the tests locally (see build.gradle and package.json) or when looking into the corresponding Bamboo build plan. -->
<!-- Lines are the main reference but a significantly lower branch percentage can indicate missing edge cases in the tests. -->
| Class/File | Branch | Line |
|------------|-------:|-----:|
| TutorialGroupResource.java | 100% | 100% |
| TutorialGroupService.java | 100% | 100% |
| tutorial-groups.service.ts | 100% | 100% |
| tutorial-groups-management.component.ts | 100% | 100% |
| tutorial-groups-management.component.ts | 100% | 100% |
| registered-students.component.ts | 73% | 93% |
| tutorial-group-form.component.ts | 69% | 91% |
| edit-tutorial-group.component.ts | 75% | 95% |
| create-tutorial-group.component.ts | 73% | 93% |
| tutorial-group-row-buttons.component.ts| 100% | 100% |


### Screenshots
![image](https://user-images.githubusercontent.com/29718932/185785808-73c75c2f-b136-442e-b743-7e6ff9cf3930.png)
![image](https://user-images.githubusercontent.com/29718932/185785841-13d7b51e-e571-4f52-85d9-a3271b445f49.png)
![image](https://user-images.githubusercontent.com/29718932/185785855-a3de1052-3b36-4e18-a20e-04efbbd705c7.png)
![image](https://user-images.githubusercontent.com/29718932/185785870-972cecab-9aa3-4421-adbd-1da58e3a12c4.png)
![image](https://user-images.githubusercontent.com/29718932/185785876-ca873be1-4cf5-404f-baae-af1272a95189.png)


### DB Changes
Note: The registration entity has been introduced because we want to allow self-registration of students later. 

![crud_model](https://user-images.githubusercontent.com/29718932/190107717-63f7f907-ac23-47dd-b563-16d7541fe1aa.png)

krusche commented on package.json:
```diff
@@ -30,6 +30,7 @@
         "@fortawesome/free-solid-svg-icons": "6.1.2",
         "@ls1intum/apollon": "2.11.2",
         "@ng-bootstrap/ng-bootstrap": "13.0.0",
+        "@ng-select/ng-select": "9.0.2",
```
do we really need that? we would like to avoid adding new dependencies if this is not 100% necessary!!!

stefanwaldhauser replied:
@krusche You are right. I found out the same can be achieved with the ngBootstrap we already have as a dependency. Removed the dependency in 7a96dfa411746e4a9daa64f7f42f750fd90957f9

krusche commented:
Database changes look ok to me

Dominik-Weinzierl commented on src/test/javascript/spec/component/tutorial-groups/tutorial-groups-management.component.spec.ts:
```diff
@@ -0,0 +1,114 @@
+
+        tutorialGroupsManagementComponentFixture.detectChanges();
+        expect(tutorialGroupsManagementComponent.tutorialGroups).toEqual([exampleTutorialGroup]);
+        expect(getAllForCourseSpy).toHaveBeenCalledWith(1);
```
Missing `toHaveBeenCalledOnce()`

stefanwaldhauser replied:
Done in d9aae13fd3c1aa34fdf4330d8e95e9c445163825

Dominik-Weinzierl commented on src/test/javascript/spec/component/tutorial-groups/tutorial-group-row-button.component.spec.ts:
```diff
@@ -0,0 +1,44 @@
+        fixture.detectChanges();
+        component.deleteTutorialGroup();
+        expect(deleteSpy).toHaveBeenCalledWith(courseId, tutorialGroup.id);
+        expect(deleteEventSpy).toHaveBeenCalled();
```
```suggestion
        expect(deleteEventSpy).toHaveBeenCalledOnce();
```

stefanwaldhauser replied:
Done in d9aae13fd3c1aa34fdf4330d8e95e9c445163825

Dominik-Weinzierl commented on src/test/javascript/spec/component/tutorial-groups/crud/create-tutorial-group.component.spec.ts:
```diff
@@ -0,0 +1,107 @@
+
+            expect(tutorialGroupCallArgument.title).toEqual(formDate.title);
+
+            expect(createStub).toHaveBeenCalledOnce();
```
Missing `toHaveBeenCalledWith()`

stefanwaldhauser replied:
Not needed as I check the call Argument above

Dominik-Weinzierl commented on src/test/javascript/spec/component/tutorial-groups/crud/create-tutorial-group.component.spec.ts:
```diff
@@ -0,0 +1,107 @@
+            expect(tutorialGroupCallArgument.title).toEqual(formDate.title);
+
+            expect(createStub).toHaveBeenCalledOnce();
+            expect(navigateSpy).toHaveBeenCalledOnce();
```
Missing `toHaveBeenCalledWith()`

stefanwaldhauser replied:
Not needed for navigate Spy. I just need to confirm navigation happens

Dominik-Weinzierl commented on src/test/javascript/spec/component/tutorial-groups/crud/edit-tutorial-group.component.spec.ts:
```diff
@@ -0,0 +1,151 @@
+
+        editTutorialGroupComponentFixture.detectChanges();
+        expect(editTutorialGroupComponent.tutorialGroup).toEqual(tutorialGroupOfResponse);
+        expect(findByIdStub).toHaveBeenCalledWith(1, 1);
```
Missing `toHaveBeenCalledOnce()`

stefanwaldhauser replied:
Done in d9aae13fd3c1aa34fdf4330d8e95e9c445163825

Dominik-Weinzierl commented on src/test/javascript/spec/component/tutorial-groups/crud/edit-tutorial-group.component.spec.ts:
```diff
@@ -0,0 +1,151 @@
+        const findByIdStub = jest.spyOn(tutorialGroupService, 'getOneOfCourse').mockReturnValue(of(findByIdResponse));
+
+        editTutorialGroupComponentFixture.detectChanges();
+        expect(findByIdStub).toHaveBeenCalledWith(1, 1);
```
Missing `toHaveBeenCalledOnce()`

Dominik-Weinzierl commented on src/test/javascript/spec/component/tutorial-groups/crud/edit-tutorial-group.component.spec.ts:
```diff
@@ -0,0 +1,151 @@
+        });
+
+        expect(updatedStub).toHaveBeenCalledOnce();
+        expect(navigateSpy).toHaveBeenCalledOnce();
```
Missing `toHaveBeenCalledWith()`

stefanwaldhauser replied:
Not needed imo as it is enough to know that navigation works.

Dominik-Weinzierl commented on src/test/javascript/spec/component/tutorial-groups/crud/tutorial-group-form.component.spec.ts:
```diff
@@ -0,0 +1,122 @@
+            providers: [
+                MockProvider(CourseManagementService, {
+                    getAllUsersInCourseGroup: (courseId: number, courseGroup: CourseGroup) => {
+                        return of(new HttpResponse({ body: [] }));
```
There are some client-style issues

Dominik-Weinzierl commented on src/test/javascript/spec/component/tutorial-groups/crud/tutorial-group-form.component.spec.ts:
```diff
@@ -0,0 +1,122 @@
+        submitButton.click();
+
+        return tutorialGroupFormComponentFixture.whenStable().then(() => {
+            expect(submitFormSpy).toHaveBeenCalledTimes(0);
```
```suggestion
            expect(submitFormEventSpy).not.toHaveBeenCalled();
```

stefanwaldhauser replied:
done in 26527db8ac4e1b72878a831b3e1e749b67b59361

Dominik-Weinzierl commented on src/test/javascript/spec/component/tutorial-groups/crud/tutorial-group-form.component.spec.ts:
```diff
@@ -0,0 +1,122 @@
+
+        return tutorialGroupFormComponentFixture.whenStable().then(() => {
+            expect(submitFormSpy).toHaveBeenCalledTimes(0);
+            expect(submitFormEventSpy).toHaveBeenCalledTimes(0);
```
```suggestion
            expect(submitFormEventSpy).not.toHaveBeenCalled();
```

stefanwaldhauser replied:
Done in 26527db8ac4e1b72878a831b3e1e749b67b59361

Dominik-Weinzierl commented on src/test/javascript/spec/component/tutorial-groups/crud/tutorial-group-form.component.spec.ts:
```diff
@@ -0,0 +1,122 @@
+        submitButton.click();
+
+        return tutorialGroupFormComponentFixture.whenStable().then(() => {
+            expect(submitFormSpy).toHaveBeenCalled();
```
```suggestion
            expect(submitFormSpy).toHaveBeenCalledOnce();
```

Dominik-Weinzierl commented on src/test/javascript/spec/component/tutorial-groups/registered-students/registered-students.component.spec.ts:
```diff
@@ -0,0 +1,130 @@
+import { User } from 'app/core/user/user.model';
+import { UserService } from 'app/core/user/user.service';
+import { CourseManagementService } from 'app/course/manage/course-management.service';
+import { Course, CourseGroup } from 'app/entities/course.model';
```
Some client style issues

Dominik-Weinzierl commented on src/test/javascript/spec/component/tutorial-groups/registered-students/registered-students.component.spec.ts:
```diff
@@ -0,0 +1,130 @@
+import { TutorialGroupsService } from 'app/course/tutorial-groups/tutorial-groups.service';
+import { RegisteredStudentsComponent } from 'app/course/tutorial-groups/registered-students/registered-students.component';
+import { ArtemisTestModule } from '../../../test.module';
+import Spy = jasmine.Spy;
```
Is this Spy needed?

stefanwaldhauser replied:
Removed

Dominik-Weinzierl commented on src/test/javascript/spec/component/tutorial-groups/registered-students/registered-students.component.spec.ts:
```diff
@@ -0,0 +1,130 @@
+            expect(comp.course).toEqual(course);
+            expect(comp.tutorialGroup).toEqual(tutorialGroup);
+            expect(comp.courseGroup).toEqual(CourseGroup.STUDENTS);
+            expect(getTutorialGroupSpy).toHaveBeenCalledTimes(1);
```
```suggestion
            expect(getTutorialGroupSpy).toHaveBeenCalledOnce();
```

stefanwaldhauser replied:
Done in 26527db8ac4e1b72878a831b3e1e749b67b59361

Dominik-Weinzierl commented on src/test/javascript/spec/component/tutorial-groups/registered-students/registered-students.component.spec.ts:
```diff
@@ -0,0 +1,130 @@
+            expect(comp.tutorialGroup).toEqual(tutorialGroup);
+            expect(comp.courseGroup).toEqual(CourseGroup.STUDENTS);
+            expect(getTutorialGroupSpy).toHaveBeenCalledTimes(1);
+            expect(comp.registeredStudents).toEqual(tutorialGroup.registeredStudents);
```
Missing `toHaveBeenCalledWith()`

stefanwaldhauser replied:
Done in 26527db8ac4e1b72878a831b3e1e749b67b59361

Dominik-Weinzierl commented on src/test/javascript/spec/component/tutorial-groups/tutorial-groups-management.component.spec.ts:
```diff
@@ -0,0 +1,114 @@
+        const sortServiceSpy = jest.spyOn(sortService, 'sortByProperty');
+
+        tutorialGroupsManagementComponent.sortRows();
+        expect(sortServiceSpy).toHaveBeenCalledWith([group1, group2], 'id', false);
```
Missing `toHaveBeenCalledOnce()`

stefanwaldhauser replied:
Done in 26527db8ac4e1b72878a831b3e1e749b67b59361

Dominik-Weinzierl commented on src/test/javascript/spec/component/shared/course-group.component.spec.ts:
```diff
@@ -34,7 +34,7 @@ jest.mock('export-to-csv', () => ({
+describe('Course Group Component', () => {
     let comp: CourseGroupComponent;
     let fixture: ComponentFixture<CourseGroupComponent>;
     let courseService: CourseManagementService;
```
there is a client code style issue

Dominik-Weinzierl commented on src/test/javascript/spec/component/course/course-group-membership.component.spec.ts:
```diff
@@ -0,0 +1,133 @@
+        });
+    });
+
+    describe('courseGroupName', () => {
+        it('should return courses studentGroupName if group is students', () => {
+            comp.courseGroup = CourseGroup.STUDENTS;
+            comp.course = { ...course };
+            comp.course.studentGroupName = 'testStudentGroupName';
+            expect(comp.courseGroupName).toBe(comp.course.studentGroupName);
+        });
+
+        it('should return courses teachingAssistantGroupName if group is tutors', () => {
+            comp.courseGroup = CourseGroup.TUTORS;
+            comp.course = { ...course };
+            comp.course.teachingAssistantGroupName = 'testTeachingAssistantGroupName';
+            expect(comp.courseGroupName).toBe(comp.course.teachingAssistantGroupName);
+        });
+
+        it('should return courses instructorGroupName if group is instructors', () => {
+            comp.courseGroup = CourseGroup.INSTRUCTORS;
+            comp.course = { ...course };
+            comp.course.instructorGroupName = 'testInstructorGroupName';
+            expect(comp.courseGroupName).toBe(comp.course.instructorGroupName);
+        });
+    });
```
I would expect parameterized tests :)

Dominik-Weinzierl commented on src/main/webapp/i18n/en/tutorialGroups.json:
```diff
@@ -0,0 +1,86 @@
+                "labelOffline": "Meeting Place",
+                "placeholderOnline": "How can the students join the online meeting?",
+                "placeholderOffline": "How can the students find the room?",
+                "maxLengthValidationError": "Location is limited to 2000 characters!"
```
Please don't write any values into the translation file.
```suggestion
                "maxLengthValidationError": "Location is limited to {{limit}} characters!"
```

Dominik-Weinzierl commented on src/main/webapp/i18n/en/tutorialGroups.json:
```diff
@@ -0,0 +1,86 @@
+                "label": "Title",
+                "placeholder": "Group 1: Monday 12:00 - 13:00",
+                "requiredValidationError": "Title is required!",
+                "maxLengthValidationError": "Title can be max 255 characters long!"
```
```suggestion
                "maxLengthValidationError": "Title can be max {{max}} characters long!"
```

Dominik-Weinzierl commented on src/main/webapp/i18n/en/tutorialGroups.json:
```diff
@@ -0,0 +1,86 @@
+        "tutorialGroupForm": {
+            "titleInput": {
+                "label": "Title",
+                "placeholder": "Group 1: Monday 12:00 - 13:00",
```
```suggestion
                "placeholder": "Group 1: Monday {{from}} - {{to}}",
```

stefanwaldhauser replied:
But this is just the placeholder?

stefanwaldhauser replied:
Done in 82a5edb5cffebb59da7db909205529cf1416d77b

Dominik-Weinzierl commented on src/main/webapp/i18n/de/tutorialGroups.json:
```diff
@@ -0,0 +1,86 @@
+                "labelOffline": "Treffpunkt",
+                "placeholderOnline": "Wie können die Studenten an der online Übung teilnehmen?",
+                "placeholderOffline": "Wo können Studenten den Raum finden?",
+                "maxLengthValidationError": "Der Ort darf maximal 2000 Zeichen lang sein!"
```
```suggestion
                "maxLengthValidationError": "Der Ort darf maximal {{limit}} Zeichen lang sein!"
```

stefanwaldhauser replied:
Done in 82a5edb5cffebb59da7db909205529cf1416d77b

Dominik-Weinzierl commented on src/main/webapp/i18n/de/tutorialGroups.json:
```diff
@@ -0,0 +1,86 @@
+            "additionalInformationInput": {
+                "label": "Zusätzliche Informationen",
+                "placeholder": "Zusätzliche Informationen zu dieser Gruppe",
+                "maxLengthValidationError": "Die zusätzlichen Informationen dürfen maximal 2000 Zeichen lang sein!"
```
```suggestion
                "maxLengthValidationError": "Die zusätzlichen Informationen dürfen maximal {{limit}} Zeichen lang sein!"
```
See the English translation file

stefanwaldhauser replied:
Done in 82a5edb5cffebb59da7db909205529cf1416d77b

Dominik-Weinzierl commented on src/main/webapp/i18n/de/tutorialGroups.json:
```diff
@@ -0,0 +1,86 @@
+        "tutorialGroupForm": {
+            "titleInput": {
+                "label": "Titel",
+                "placeholder": "Gruppe 1: Montag 12:00 - 13:00",
```
```suggestion
                "placeholder": "Gruppe 1: Montag {{from}} - {{to}}",
```

stefanwaldhauser replied:
Done in 82a5edb5cffebb59da7db909205529cf1416d77b

Dominik-Weinzierl commented on src/main/webapp/i18n/de/tutorialGroups.json:
```diff
@@ -0,0 +1,86 @@
+                "label": "Titel",
+                "placeholder": "Gruppe 1: Montag 12:00 - 13:00",
+                "requiredValidationError": "Eine Übungsgruppe muss einen Titel haben!",
+                "maxLengthValidationError": "Der Titel darf maximal 255 Zeichen lang sein!"
```
```suggestion
                "maxLengthValidationError": "Der Titel darf maximal {{max}} Zeichen lang sein!"
```

stefanwaldhauser replied:
Done in 82a5edb5cffebb59da7db909205529cf1416d77b

Dominik-Weinzierl reviewed (dismissed):
Some suggestions and change requests :+1: i did not check the entire code, so this counts not as code review :) 

Please avoid hard coding values in the translation file.

canberkanar reviewed (changes requested):
Please relocate the "submit" button to the right-end side.  Also, I think "save" is more relevant here. Other than that, good job! 👍🏻 

![image](https://user-images.githubusercontent.com/67344158/186655556-89927d59-bbd2-49d4-8303-04c79bb1f1b5.png)

ge65cer reviewed (changes requested):
Tested in testing session on TS1. When adding a user to a tutorial group and then trying to delete the user in user management, the server responds with error 500 due to a constraint violation. Removing the user from the tutorial group and then deleting the user works, though.

akesfeden reviewed (commented):
Tested on TS1 during testing session, functionality works as expected. I can approve after the suggestions related to submit/save button is implemented.

manuelmanso commented:
> Tested in testing session on TS1. When adding a user to a tutorial group and then trying to delete the user in user management, the server responds with error 500 due to a constraint violation. Removing the user from the tutorial group and then deleting the user works, though.

This is related to this issue: https://github.com/ls1intum/Artemis/issues/2353 , not unique to this PR, this PR just adds an extra relation that isn't deleted

ge65cer commented:
@manuelmanso No, as I said when removing the user from the tutor group I was able to delete the user successfully.

stefanwaldhauser commented:
> Please relocate the "submit" button to the right-end side. Also, I think "save" is more relevant here. Other than that, good job! 👍🏻
> 
> ![image](https://user-images.githubusercontent.com/67344158/186655556-89927d59-bbd2-49d4-8303-04c79bb1f1b5.png)

I just checked and we have it on the left side for example look at course creation. But I changed the text depending on edit or create in 3379d5907c6ca00a6c980ad5a29c82b5bfbf3c5c

akesfeden reviewed (dismissed):
Re-tested on TS1 after feedback is implemented, CRUD still works, student addition and deletion also works. Existing functionality of adding students and instructors to a course is also working as before.

Hialus commented on src/main/java/de/tum/in/www1/artemis/web/rest/tutorialgroups/TutorialGroupResource.java:
```diff
@@ -0,0 +1,305 @@
+    }
+
+    private void checkTitleIsUnique(TutorialGroup tutorialGroup) {
+        if (tutorialGroupRepository.findAllByCourseId(tutorialGroup.getCourse().getId()).stream().map(TutorialGroup::getTitle)
```
This could be moved into a special SQL query, which would greatly improve the performance, as you would not have to fetch all existing tutor groups of the course.

stefanwaldhauser replied:
Very good idea! Done in  f2af95843922f41c3d4ef62efd17aa3f02016222 :)

Hialus commented on src/main/java/de/tum/in/www1/artemis/web/rest/tutorialgroups/TutorialGroupResource.java:
```diff
@@ -0,0 +1,305 @@
+    @PutMapping("/courses/{courseId}/tutorial-groups")
+    @PreAuthorize("hasRole('INSTRUCTOR')")
+    @FeatureToggle(Feature.TutorialGroups)
+    public ResponseEntity<TutorialGroup> update(@PathVariable Long courseId, @RequestBody @Valid TutorialGroup tutorialGroup) {
```
This update endpoint is currently according to Artemis standard. 
We normally use a path like `PUT /api/entity/{entityId}` and then take the entityId as a parameter. We then check if the entityId equals the id of the request body.

stefanwaldhauser replied:
Done n 3b4418b238b4f4c15319f4688fc1d24d6f2897a3

Hialus commented on src/main/java/de/tum/in/www1/artemis/web/rest/tutorialgroups/TutorialGroupResource.java:
```diff
@@ -0,0 +1,305 @@
+            throw new EntityNotFoundException("User", studentLogin);
+        }
+        tutorialGroupService.deregisterStudent(studentToDeregister.get(), tutorialGroupFromDatabase);
+        return ResponseEntity.ok().body(null);
```
Usually we return `ResponseEntity.noContent()` here.

stefanwaldhauser replied:
Done in 9eca40ec5100cab4110ea542c7bfc5de7a6f440e

Hialus commented on src/main/java/de/tum/in/www1/artemis/web/rest/tutorialgroups/TutorialGroupResource.java:
```diff
@@ -0,0 +1,305 @@
+        checkTutorialCourseIdMatchesPathId(courseId, tutorialGroupFromDatabase);
+        authorizationCheckService.checkHasAtLeastRoleInCourseElseThrow(Role.INSTRUCTOR, tutorialGroupFromDatabase.getCourse(), null);
+        tutorialGroupRepository.deleteById(tutorialGroupFromDatabase.getId());
+        return ResponseEntity.ok().headers(HeaderUtil.createEntityDeletionAlert(applicationName, true, ENTITY_NAME, tutorialGroupFromDatabase.getTitle())).build();
```
Usually we return ResponseEntity.noContent() here.

stefanwaldhauser replied:
Done in 9eca40ec5100cab4110ea542c7bfc5de7a6f440e

Hialus commented on src/main/java/de/tum/in/www1/artemis/web/rest/tutorialgroups/TutorialGroupResource.java:
```diff
@@ -0,0 +1,305 @@
+            throw new EntityNotFoundException("User", studentLogin);
+        }
+        tutorialGroupService.registerStudent(studentToRegister.get(), tutorialGroupFromDatabase);
+        return ResponseEntity.ok().body(null);
```
`ResponseEntity.noContent()` might also make sense here.

stefanwaldhauser replied:
Done in 94342a84726518898c07e0ced71a30fc0c478410

Hialus commented on src/main/java/de/tum/in/www1/artemis/web/rest/tutorialgroups/TutorialGroupResource.java:
```diff
@@ -0,0 +1,305 @@
+        var tutorialGroupFromDatabase = this.tutorialGroupRepository.findByIdWithTeachingAssistantAndRegistrationsElseThrow(tutorialGroupId);
+        checkTutorialCourseIdMatchesPathId(courseId, tutorialGroupFromDatabase);
+        authorizationCheckService.checkHasAtLeastRoleInCourseElseThrow(Role.INSTRUCTOR, tutorialGroupFromDatabase.getCourse(), null);
+        List<StudentDTO> notFoundStudentsDtos = tutorialGroupService.registerMultipleStudents(courseId, tutorialGroupId, studentDtos);
```
Same as in the service.
```suggestion
        List<StudentDTO> notFoundStudentDtos = tutorialGroupService.registerMultipleStudents(courseId, tutorialGroupId, studentDtos);
```

stefanwaldhauser replied:
Done in 3476671f7961517424bc126fb63cb0ad757d0dc9

Hialus commented on src/main/java/de/tum/in/www1/artemis/domain/tutorialgroups/TutorialGroupRegistration.java:
```diff
@@ -0,0 +1,68 @@
+    private TutorialGroupRegistrationType type;
+
+    public TutorialGroupRegistration() {
+        // Empty constructor
```
For consistency: (Also see my other comment)
```suggestion
        // Empty constructor needed for Jackson.
```

Hialus reviewed (changes requested):
Reviewed the server code and left a few suggestions.

Also a general question out of curiosity: Why did you decide to rename `TutorGroup` to `TutorialGroup`?

The server tests currently fail on this PR, which is why I did not review them. Please fix this.

stefanwaldhauser commented on src/main/java/de/tum/in/www1/artemis/domain/tutorialgroups/TutorialGroupRegistration.java:
```diff
@@ -0,0 +1,68 @@
+public class TutorialGroupRegistration extends DomainObject {
+
+    @ManyToOne
+    @JoinColumn(name = "student_id")
```
I like to be explicit with the columns. I see no harm in specifying them.

stefanwaldhauser commented on src/main/java/de/tum/in/www1/artemis/domain/User.java:
```diff
@@ -145,6 +146,10 @@ public class User extends AbstractAuditingEntity implements Participant {
     private Set<Organization> organizations = new HashSet<>();
 
+    @OneToMany(mappedBy = "student", fetch = FetchType.LAZY, cascade = CascadeType.REMOVE, orphanRemoval = true)
+    @Cache(usage = CacheConcurrencyStrategy.NONSTRICT_READ_WRITE)
```
Are you quite sure of this? I looked at other properties where we also use it on the collection and the hibernate documentation says the following: https://docs.jboss.org/hibernate/core/3.6/reference/de-DE/html/performance.html
![image](https://user-images.githubusercontent.com/29718932/187796078-c0daedf7-0544-4924-9271-ec16ad3e8047.png)

Hialus replied:
I think you are correct. Haven't used this in quite a while, so I forgot about this.

stefanwaldhauser commented on src/main/java/de/tum/in/www1/artemis/service/TutorialGroupService.java:
```diff
@@ -0,0 +1,95 @@
+                notFoundStudentsDTOs.add(studentDto);
+            }
+            else {
+                registerStudent(optionalStudent.get(), tutorialGroup);
```
Registration happens very early just in the beginning of the semester, and rarely throughout the course. I will keep this solution therefore.

ge65cer replied:
I agree with @Hialus

Hialus replied:
Just because it happens rarely doesn't mean you have to write it purposefully inefficiently. Unless the change is too complex, we should always take the more performant approach (especially when it comes to the DB).
In this case this would only be about extracting part of the registerStudent method into a helper method and collecting them here. This should only take about 5 minutes and decreases the amount of DB calls by around a factor of 20, so you should definitely do this.

stefanwaldhauser replied:
Done in 1f00c8c9e9259eacd45b6c44bbb9590de5186fde

ge65cer commented on src/main/java/de/tum/in/www1/artemis/service/user/UserService.java:
```diff
@@ -434,10 +443,19 @@ protected void deleteUser(User user) {
         studentScoreRepository.deleteAllByUser(user);
         exerciseHintActivationRepository.deleteAllByUser(user);
 
+        tutorialGroupRegistrationRepository.deleteAllByStudent(user);
```
Why do you do this manually here instead of using a cascade annotation?

stefanwaldhauser replied:
I tried it and it did not work with the cascade annotation alone.

Hialus commented on src/main/java/de/tum/in/www1/artemis/domain/tutorialgroups/TutorialGroup.java:
```diff
@@ -0,0 +1,156 @@
+    private Set<TutorialGroupRegistration> registrations = new HashSet<>();
+
+    public TutorialGroup() {
+        // Empty constructor needed for Jackson.
```
I personally don't think that "used in test" is a good reason for doing something. 
If you want to create a TutorialGroup in the tests easier, then you would normally write a helper method in e.g. DatabaseUtilService or ModelFactory.

stefanwaldhauser commented on src/main/java/de/tum/in/www1/artemis/domain/tutorialgroups/TutorialGroup.java:
```diff
@@ -0,0 +1,156 @@
+
+    @Column(name = "title")
+    @Size(max = 256)
+    @NotBlank()
```
Done in 46defb986f645804d6ad9e8ca7b46683048e4807

stefanwaldhauser commented on src/main/java/de/tum/in/www1/artemis/service/TutorialGroupService.java:
```diff
@@ -0,0 +1,95 @@
+        for (var studentDto : studentDTOs) {
+            var registrationNumber = studentDto.getRegistrationNumber();
+            var login = studentDto.getLogin();
+            Optional<User> optionalStudent = userService.findUserAndAddToCourse(registrationNumber, course.getStudentGroupName(), Role.STUDENT, login);
```
Done in 1f00c8c9e9259eacd45b6c44bbb9590de5186fde

stefanwaldhauser replied:
I also restricted the user search to only students of the course here: c41b192819acc35b038fd32c18c9238b297c9a7f

stefanwaldhauser commented on src/main/java/de/tum/in/www1/artemis/service/TutorialGroupService.java:
```diff
@@ -0,0 +1,95 @@
+        var course = this.courseRepository.findByIdElseThrow(courseId);
+        var tutorialGroup = tutorialGroupRepository.findByIdWithTeachingAssistantAndRegistrationsElseThrow(tutorialGroupId);
+
+        List<StudentDTO> notFoundStudentsDTOs = new ArrayList<>();
```
Renamed in a891d08cabd901b67448dd2f661b86f47930525b

stefanwaldhauser commented on src/main/java/de/tum/in/www1/artemis/service/user/UserService.java:
```diff
@@ -434,10 +443,19 @@ protected void deleteUser(User user) {
         // 9) Delete the submissionVersion / remove the user from the submissionVersion
         // 10) Delete the tutor participation
+        // 11) All tutorial group registrations of the student
+        // 12) All tutorial groups taught by the user
```
Done in 842ef752ea10821b0c8906b28deb7535b4f88650

Hialus commented on src/main/java/de/tum/in/www1/artemis/service/TutorialGroupService.java:
```diff
@@ -0,0 +1,114 @@
+        var registrationNumber = studentDto.getRegistrationNumber();
+        var login = studentDto.getLogin();
+
+        // try to find the user by login
```
```suggestion
        // try to find the user by registration number
```

Hialus commented on src/main/java/de/tum/in/www1/artemis/service/TutorialGroupService.java:
```diff
@@ -0,0 +1,114 @@
+        }
+        // try to find the student by login
+        var studentByLogin = userRepository.findUserWithGroupsAndAuthoritiesByLogin(login);
+        if (studentByLogin.isPresent()) {
```
Optional: This if statement is a duplicate of the one above. You could combine them.

stefanwaldhauser replied:
Good idea! I simplified the code logic in: 1ec292a7ac629d1af56794ca12d4627f44a25b27

Hialus commented on src/main/java/de/tum/in/www1/artemis/web/rest/tutorialgroups/TutorialGroupResource.java:
```diff
@@ -0,0 +1,279 @@
+    @FeatureToggle(Feature.TutorialGroups)
+    public ResponseEntity<String> getTitle(@PathVariable Long tutorialGroupId) {
+        final var title = tutorialGroupRepository.getTutorialGroupTitle(tutorialGroupId);
+        return title == null ? ResponseEntity.notFound().build() : ResponseEntity.ok(title);
```
`ResponseEntity.notFound()` does not offer a very descriptive error message, so we usually use an `EntityNotFoundException`, which gives a bit more detail.
But you can consider this an optional change, as it shouldn't really matter.

stefanwaldhauser replied:
Done in 5a454831567eb4226566e1e0bae347a68c357cdc

Hialus commented on src/main/java/de/tum/in/www1/artemis/web/rest/tutorialgroups/TutorialGroupResource.java:
```diff
@@ -0,0 +1,279 @@
+
+        authorizationCheckService.checkHasAtLeastRoleInCourseElseThrow(Role.INSTRUCTOR, course, user);
+
+        // ToDo: Optimization Idea: Do not send all registered student information but just the number in a DTO
```
I agree. Though in my testing with 2000 users the endpoint actually performed better than expected (<1s), so this might not be that high priority.

stefanwaldhauser replied:
Will do it in further optimization follow up

Hialus commented on src/main/java/de/tum/in/www1/artemis/web/rest/tutorialgroups/TutorialGroupResource.java:
```diff
@@ -0,0 +1,279 @@
+        var tutorialGroupFromDatabase = this.tutorialGroupRepository.findByIdElseThrow(tutorialGroupId);
+        authorizationCheckService.checkHasAtLeastRoleInCourseElseThrow(Role.INSTRUCTOR, tutorialGroupFromDatabase.getCourse(), null);
+
+        Optional<User> studentToDeregister = userRepository.findOneWithGroupsAndAuthoritiesByLogin(studentLogin);
```
You could use `userRepository.getUserWithGroupsAndAuthorities()`, as this already has the isEmpty check with the correct `EntityNotFoundException` included.

stefanwaldhauser replied:
Good idea! Done in 282a84a2a83ea0d17d0d693e02b3114c978652d5

Hialus commented on src/main/java/de/tum/in/www1/artemis/web/rest/tutorialgroups/TutorialGroupResource.java:
```diff
@@ -0,0 +1,279 @@
+        var tutorialGroupFromDatabase = this.tutorialGroupRepository.findByIdElseThrow(tutorialGroupId);
+        authorizationCheckService.checkHasAtLeastRoleInCourseElseThrow(Role.INSTRUCTOR, tutorialGroupFromDatabase.getCourse(), null);
+
+        Optional<User> userToRegister = userRepository.findOneWithGroupsAndAuthoritiesByLogin(studentLogin);
```
Same here.

stefanwaldhauser replied:
Good idea! Done in 282a84a2a83ea0d17d0d693e02b3114c978652d5

Hialus commented on src/main/java/de/tum/in/www1/artemis/web/rest/tutorialgroups/TutorialGroupResource.java:
```diff
@@ -0,0 +1,279 @@
+    @PutMapping("/tutorial-groups")
+    @PreAuthorize("hasRole('INSTRUCTOR')")
+    @FeatureToggle(Feature.TutorialGroups)
+    public ResponseEntity<TutorialGroup> update(@RequestBody @Valid TutorialGroup tutorialGroup) {
```
I think you misunderstood me here.
I did not mean that you should remove the `/course/{course-id}` here, but that you should also include the `{tutorialGroupId}` in the URL, so it would be like this:
```java
@PutMapping("programming-exercises/{exerciseId}/exercise-hints/{exerciseHintId}")
    @PreAuthorize("hasRole('EDITOR')")
    public ResponseEntity<ExerciseHint> updateExerciseHint(@RequestBody ExerciseHint exerciseHint, @PathVariable Long exerciseHintId, @PathVariable Long exerciseId) {
```

Sorry if I wasn't clear enough here. The course part is correct, but you could also leave it out if you want (Though you should ask Stephan if he is fine with you leaving it out). You also currently aren't consistent in this file.

stefanwaldhauser replied:
@Hialus I left it in where I need the information. For example when creating a tutorial group I want to know to which course. I could get this also from the @RequestBody but I find it a little bit clearer.

I like this simplified way now where I only put the id in the URL I really need which means I do not have to pass around ids that much in the components in the client.

What is your opinion @krusche ?

ge65cer replied:
I agree with @Hialus, for a PUT request that updates a single entity you'd normally want to have the id as part of the URL.
And to make it consistent with other resources, we normally include the course id as well.

stefanwaldhauser replied:
I made all endpoints consistent in c926db4b405b0d47b989056509025ea28d1098c5

Hialus reviewed (changes requested):
Thanks for including my suggestions. I went through the code in more detail again and found mostly some nitpicks but also a couple more important changes.

Also regarding your UML diagram:
1. It is hard to read in dark mode 😉 
2. There is no connection between `TutorialGroupRegistration` and `RegistrationType`; instead, the registration has a "type" attribute. You should replace the attribute with a connection imo.
<img width="271" alt="grafik" src="https://user-images.githubusercontent.com/18218285/188757718-c817fbe0-c850-499f-886d-4e5b31bc28a8.png">
3. The labels for the user a misplaced by a bit
<img width="169" alt="grafik" src="https://user-images.githubusercontent.com/18218285/188757883-99d1fdcb-f707-48d3-85a8-f7581ab07e22.png">
4. You use it to highlight the DB changes, despite the diagram actually being an AOM. Just FYI, as a database model is slightly different. You don't have to change it for this PR though but should probably do so for your thesis.

>
Also, I like your detailed screenshots. 👍🏻

stefanwaldhauser commented on src/main/java/de/tum/in/www1/artemis/repository/UserRepository.java:
```diff
@@ -246,6 +259,19 @@ default Page<UserDTO> searchAllUsersByLoginOrName(Pageable pageable, String logi
+     * @param groupName   Group name to search in
+     * @return all users matching search criteria in the group
+     */
+    default Page<UserDTO> searchAllUsersByLoginOrNameInGroup(Pageable pageable, String loginOrName, String groupName) {
```
Very good idea! I changed the method name and javadoc here: 4efbba79f4d37d39b6886b0d99652e31f5789f3e

stefanwaldhauser commented on src/main/java/de/tum/in/www1/artemis/web/rest/tutorialgroups/TutorialGroupResource.java:
```diff
@@ -0,0 +1,279 @@
+    }
+
+    private void checkTitleIsUnique(TutorialGroup tutorialGroup) {
+        if (tutorialGroupRepository.findByTitle(tutorialGroup.getTitle()).isPresent()) {
```
Amazing catch @Hialus 👍. Thank you! How could I miss that :) Thank you! Fixed in 69dd27768bc8fefa92683306b428faf72c9b6eda

stefanwaldhauser commented on src/main/java/de/tum/in/www1/artemis/repository/tutorialgroups/TutorialGroupRepository.java:
```diff
@@ -0,0 +1,66 @@
+            ORDER BY tutorialGroup.title""")
+    Set<TutorialGroup> findAllByCourseId(@Param("courseId") Long courseId);
+
+    Optional<TutorialGroup> findByTitle(String title);
```
Very good idea! Done in 87468773564ad21032eb6bf1bf3558022e0d862a

ge65cer commented on src/main/java/de/tum/in/www1/artemis/web/rest/tutorialgroups/TutorialGroupResource.java:
```diff
@@ -0,0 +1,279 @@
+     * @param tutorialGroupId the id of the tutorial group to delete
+     * @return the ResponseEntity with status 204 (NO_CONTENT)
+     */
+    @DeleteMapping("/tutorial-groups/{tutorialGroupId}")
```
Here you might want to add the course as path to make it consistent with other endpoints.

stefanwaldhauser replied:
Done in c926db4b405b0d47b989056509025ea28d1098c5

ge65cer commented on src/main/java/de/tum/in/www1/artemis/web/rest/tutorialgroups/TutorialGroupResource.java:
```diff
@@ -0,0 +1,279 @@
+    @DeleteMapping("/tutorial-groups/{tutorialGroupId}/deregister/{studentLogin:" + Constants.LOGIN_REGEX + "}")
+    @PreAuthorize("hasRole('INSTRUCTOR')")
+    @FeatureToggle(Feature.TutorialGroups)
+    public ResponseEntity<Void> deregisterStudent(@PathVariable Long tutorialGroupId, @PathVariable String studentLogin) {
```
Again, here you might want to add the course as path to make it consistent with other endpoints.

stefanwaldhauser replied:
Done in c926db4b405b0d47b989056509025ea28d1098c5

ge65cer commented on src/main/java/de/tum/in/www1/artemis/web/rest/tutorialgroups/TutorialGroupResource.java:
```diff
@@ -0,0 +1,279 @@
+    @PostMapping("/tutorial-groups/{tutorialGroupId}/register/{studentLogin:" + Constants.LOGIN_REGEX + "}")
+    @PreAuthorize("hasRole('INSTRUCTOR')")
+    @FeatureToggle(Feature.TutorialGroups)
+    public ResponseEntity<Void> registerStudent(@PathVariable Long tutorialGroupId, @PathVariable String studentLogin) {
```
Again, here you might want to add the course as path to make it consistent with other endpoints.

stefanwaldhauser replied:
Done in c926db4b405b0d47b989056509025ea28d1098c5

ge65cer commented on src/main/java/de/tum/in/www1/artemis/web/rest/tutorialgroups/TutorialGroupResource.java:
```diff
@@ -0,0 +1,279 @@
+    @PostMapping("/tutorial-groups/{tutorialGroupId}/register-multiple")
+    @PreAuthorize("hasRole('INSTRUCTOR')")
+    @FeatureToggle(Feature.TutorialGroups)
+    public ResponseEntity<Set<StudentDTO>> registerMultipleStudentsToTutorialGroup(@PathVariable Long tutorialGroupId, @RequestBody Set<StudentDTO> studentDtos) {
```
Do you even need the singular endpoints (line 195 and 218)? I could imagine you could use this endpoint to register one student as well. That would mean less code to maintain.

ge65cer commented on src/main/resources/config/liquibase/changelog/20220731000000_changelog.xml:
```diff
@@ -0,0 +1,64 @@
+
+        <createTable tableName="tutorial_group_registration">
+            <column autoIncrement="true" name="id" type="BIGINT">
+                <constraints nullable="false" primaryKey="true" primaryKeyName="tutorial_groupPK"/>
```
Does you relation need a primary key or could the tuple (tutorial_group_id, student_id) act as one?

stefanwaldhauser replied:
We might support in the future that a student could self register to a tutorial group and be registered by an instructor. So to keep this possibility open, I would rather not make this uniqueness restriction for now.

ge65cer replied:
How would that feature influence the primary key in this table? Could you elaborate a little more? 🙂

stefanwaldhauser replied:
Well if I make tutorial_group_id and student_id the primary key only one registration could exist for one student and one tutorial group if I am not mistaken.

Hialus replied:
This is actually a really good catch, and I overlooked this. As it currently stands, this would cause issues further down the road. See my other comment.

Hialus commented on src/main/java/de/tum/in/www1/artemis/repository/tutorialgroups/TutorialGroupRegistrationRepository.java:
```diff
@@ -0,0 +1,22 @@
+@Repository
+public interface TutorialGroupRegistrationRepository extends JpaRepository<TutorialGroupRegistration, Long> {
+
+    Optional<TutorialGroupRegistration> findTutorialGroupRegistrationByTutorialGroupAndStudent(TutorialGroup tutorialGroup, User student);
```
Regarding my comment in the migration file:
If you allow multiple entries for a single (student, tutorialGroup) pair, then this query will not work reliably. If there is more than one entry for such a pair it will throw this exception:
`javax.persistence.NonUniqueResultException: query did not return a unique result: 2`.

Considering that this PR does not include the self-registration functionality, this should not be an issue currently, but once you implement it and allow multiple entries to exist, then it will become a problem.

Considering this you don't have to fix this issue in this PR, but considering that it might require bigger changes you should consider it.

stefanwaldhauser replied:
Good catch! I noted this down and will consider this once we implement self registration. Thank you 👍

Hialus reviewed (changes requested):
And the client tests are currently failing. Please fix this.

Hialus commented on src/main/java/de/tum/in/www1/artemis/web/rest/tutorialgroups/TutorialGroupResource.java:
```diff
@@ -0,0 +1,322 @@
+     * @param tutorialGroupId the id of the tutorial group
+     * @return ResponseEntity with status 200 (OK) and with body containing the title of the tutorial group
+     */
+    @GetMapping("/tutorial-groups/{tutorialGroupId}/title")
```
Unlike all other endpoints, this one does not include the `/courses/{courseId}` part. For consistency, you should consider adding it.

stefanwaldhauser replied:
@Hialus this does not work as the service resolving the breadcrumbs expects this structure of the endpoint.

stefanwaldhauser replied:
```
    private fetchTitle(type: EntityType, ids: number[]): void {
        let resourceUrl = 'api/';
        switch (type) {
            case EntityType.COURSE:
                resourceUrl += 'courses';
                break;
            case EntityType.EXERCISE:
                resourceUrl += 'exercises';
                break;
            case EntityType.LECTURE:
                resourceUrl += 'lectures';
                break;
            case EntityType.HINT:
                resourceUrl += `programming-exercises/${ids[1]}/exercise-hints`;
                break;
            case EntityType.DIAGRAM:
                resourceUrl += 'apollon-diagrams';
                break;
            case EntityType.EXAM:
                resourceUrl += 'exams';
                break;
            case EntityType.ORGANIZATION:
                resourceUrl += 'organizations';
                break;
        }

        this.http.get(`${SERVER_API_URL}${resourceUrl}/${ids[0]}/title`, { observe: 'response', responseType: 'text' }).subscribe((response: HttpResponse<string>) => {
            if (response.body) {
                this.setTitle(type, ids, response.body);
            }
        });
    }
```
In entity-title.service.ts

Hialus replied:
From what I can see, this would not prevent using the path `courses/{id}/tutorial-groups/{id}`, as the exercise hints also use such a format.
But I am not that good regarding the client code, so I might be from here.

stefanwaldhauser commented:
I removed the `location` property from tutorial group as it makes sense in another entity that will be introduced in the session management follow up :)

Hialus reviewed (approved):
The server code looks good to me now. Great job! 🚀 
I only found one very minor nitpick, that doesn't matter that much.

stefanwaldhauser commented on src/main/java/de/tum/in/www1/artemis/repository/tutorialgroups/TutorialGroupRepository.java:
```diff
@@ -0,0 +1,72 @@
+import de.tum.in.www1.artemis.web.rest.errors.EntityNotFoundException;
+
+@Repository
+public interface TutorialGroupRepository extends JpaRepository<TutorialGroup, Long> {
```
Done in a0b8ae3f06b134463d6787c751a8e6c80ab78492

stefanwaldhauser commented:
> Tested on TS3
> 
> * You can only select tutors. As the groups are not that strictly distributed, one should be able to assign Tutors, Editors, or Instructors

> * It's not possible to set the group to Online despite the option. Neither on create nor on edit


> * I think the Button for the Edit form should rather be called "Save" with the appropriate icon (see e.g. the course form)


> * I think the switches are hard to operate, I'd advise introducing colors or using a select instead
- I had a select originally but in a discussion with @krusche we decided that switches are more appropriate. But nevertheless I have now introduced colors in: 


> * I'm missing the room for offline and the link for online. Or is this meant to go in the additional information? I'd advise against that as this could prevent sharing the information automatically when groups get merged, etc.
- As I have written in a previous comment above, I have moved this information into a different entity introduced in a follow up.


> * You can add more students to the group than the group has capacity. While this could make sense in certain cases, it's unintuitive right now. At least a warning should be present
- We have not finally discussed how we want to restrict registration right now. Consider this part not done yet. Therefore I will not restrict it right now in this PR. Of course before we switch that feature "live" some decision has to be made here.

> * Does it make sense that one student can be registered in multiple groups?
- We want to allow students to be "interested" in more than one group, so they can get access for example to discussion boards of multiple groups. Thats why I do not restrict it right now.

> * You should extend the header of the table. Right now it's short one column
- As discussed with @krusche I removed certain headers like additional Information in the table. These will be shown in a detail page similiar to other entities


> Considering your PR description, I would say that this PR is insufficient, and you should work on that feedback. In case you just wanted to create a basis and plan to include certain inconsistencies in another PR, let me know and I check again.

julian-christl commented:
The fixes work as intended.

As already suggested, you should define your scope of the PR more cleary but the way you describe it in the comment, it's sufficient now.

> As discussed with @krusche I removed certain headers like additional Information in the table. These will be shown in a detail page similiar to other entities

I'm talking about something different:

Table of the tutorial groups management missing one column in the header => line stops at before the buttons
![](https://gyazo.com/93ce2fcacc5336710a9b7942b5dd9e6c/raw)

Table of the text exercises misses no column in the header => line goes through to the end.
![](https://gyazo.com/474ded7586364cdacc2c62f4f0e93098/raw)

stefanwaldhauser commented:
> The fixes work as intended.
> 
> As already suggested, you should define your scope of the PR more cleary but the way you describe it in the comment, it's sufficient now.
> 
> > As discussed with @krusche I removed certain headers like additional Information in the table. These will be shown in a detail page similiar to other entities
> 
> I'm talking about something different:
> 
> Table of the tutorial groups management missing one column in the header => line stops at before the buttons ![](https://camo.githubusercontent.com/d69716ff78a53213d3de72e23b1e0db33ee82c342579f1ed8a6f30e52baf91aa/68747470733a2f2f6779617a6f2e636f6d2f39336365326663616363353333363731306139623739343262356464396536632f726177)
> 
> Table of the text exercises misses no column in the header => line goes through to the end. ![](https://camo.githubusercontent.com/f96672bea4206b3985274627e8cfb400d0b643635a7ed9b8841dfecdfee7b061/68747470733a2f2f6779617a6f2e636f6d2f34373464656437353836333634636461636332633632663466306539333039382f726177)

Fixed in 1aa35fb0738238fe11fbaa81a0c6c1306c6b560b

julian-christl reviewed (dismissed):
Tested on TS3
Considering the scope of the PR, it works well.

I didn't retest the last fix, I cross-checked it with the code as it was only one HTML tag missing.

akesfeden reviewed (approved):
Tested on TS2, CRUD functionality works as expected and the UI looks nice, good work!

simon-hng commented on src/main/java/de/tum/in/www1/artemis/web/rest/tutorialgroups/TutorialGroupResource.java:
```diff
@@ -0,0 +1,323 @@
+    }
+
+    private static void overrideValues(TutorialGroup sourceTutorialGroup, TutorialGroup originalTutorialGroup) {
+        originalTutorialGroup.setTitle(sourceTutorialGroup.getTitle());
```
Make those setter return `Tutorgroup` so you can write
```java
originalTutorialGroup
  .setTeachingAssisteant(...)
  .setAdditionalInformation(...)...
```

stefanwaldhauser replied:
I would rather not do this. I read up on it online and apparently this causes problems as it fails the JavaBean Specification. 
```
it violates the JavaBean spec, so it will break with various tools and libraries, especially JSP builders and Spring.
```
See:  https://stackoverflow.com/questions/1345001/is-it-bad-practice-to-make-a-setter-return-this

simon-hng commented on src/main/webapp/app/course/tutorial-groups/tutorial-groups-instructor-view/tutorial-groups/crud/create-tutorial-group/create-tutorial-group.component.ts:
```diff
@@ -0,0 +1,61 @@
+
+        const { title, teachingAssistant, additionalInformation, capacity, isOnline, language, campus } = formData;
+
+        this.tutorialGroupToCreate.title = title;
```
using setter and returning TutorialGroup for each could also benefit you here, similar to the comment before this.

(Not writing `this.tutorialGroupToCreate` 5 times helps with not straining your arms ;))

stefanwaldhauser replied:
Could be a nice optimisation for the future but I want to get this PR done now. This does not seem critical.

simon-hng replied:
No java bean specs in typescript. Also why don't we use constructors?

akesfeden reviewed (approved):
Reapprove after translation string refactoring and test fix.

julian-christl reviewed (approved):
reapprove, no functional changes
